ROOT and CUDADataFormats/Common/interface/Product.h

Working in studies of the GPU/CPU difference , I spot this error of ROOT with the abs function:
```
root -l /afs/cern.ch/work/s/sdonato/public/GPU_fluctuation_study/output_HLT_GPU_CPU_GPU2.root -b -q -e "Events->Scan(\"abs\(1\)\")"
```
the log is here [logError.txt](https://github.com/cms-sw/cmssw/files/8233562/logError.txt) .

@dpiparo has identificated the problem with `CUDADataFormats/Common/interface/Product.h`

```
[sdonato@lxplus764 src]$ root -b
root [0] #include "/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/CUDADataFormats/Common/interface/Product.h"
In file included from ROOT_prompt_0:1:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/CUDADataFormats/Common/interface/Product.h:6:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/CUDADataFormats/Common/interface/ProductBase.h:7:
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm.lock-fc52a3ea: Read-only file system [-Rmodule-build]
#include <cuda_runtime.h>
         ^
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: building module 'cuda' as '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm': 'Read-only file system'
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: finished building module 'cuda' [-Rmodule-build]
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_0_pre6/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
```

or 
```
[sdonato@lxplus764 src]$ root -l 
root [0] #include <cuda_runtime.h>
ROOT_prompt_0:1:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm.lock-30779e7c: Read-only file system [-Rmodule-build]
#include <cuda_runtime.h>
         ^
ROOT_prompt_0:1:10: remark: building module 'cuda' as '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/lib/cuda.pcm': 'Read-only file system'
ROOT_prompt_0:1:10: remark: finished building module 'cuda' [-Rmodule-build]
ROOT_prompt_0:1:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
root [1] 
```

--- a/CUDADataFormats/CaloCommon/src/classes_def.xml
+++ b/CUDADataFormats/CaloCommon/src/classes_def.xml
@@ -2,7 +2,7 @@
     <enum name="calo::common::tags::Ptr"/>
     <enum name="calo::common::tags::Vec"/>
     <enum name="calo::common::tags::DevPtr"/>
-    <class name="calo::common::AddSize<calo::common::tags::Ptr>"/>
-    <class name="calo::common::AddSize<calo::common::tags::Vec>"/>
-    <class name="calo::common::AddSize<calo::common::tags::DevPtr>"/>
+    <class name="calo::common::AddSize<calo::common::tags::Ptr>" persistent="false"/>
+    <class name="calo::common::AddSize<calo::common::tags::Vec>" persistent="false"/>
+    <class name="calo::common::AddSize<calo::common::tags::DevPtr>" persistent="false"/>
 </lcgdict>

--- a/CUDADataFormats/HcalDigi/src/classes_def.xml
+++ b/CUDADataFormats/HcalDigi/src/classes_def.xml
@@ -1,14 +1,14 @@
 <lcgdict>
-    <class name="hcal::DigiCollectionBase<calo::common::VecStoragePolicy<std::allocator>>" />
-    <class name="hcal::DigiCollectionBase<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" />
+    <class name="hcal::DigiCollectionBase<calo::common::VecStoragePolicy<std::allocator>>" persistent="false" />
+    <class name="hcal::DigiCollectionBase<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" persistent="false" />
 
-    <class name="hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<std::allocator>>" />
-    <class name="hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<std::allocator>>" />
-    <class name="hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<std::allocator>>" />
+    <class name="hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<std::allocator>>" persistent="false" />
+    <class name="hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<std::allocator>>" persistent="false" />
+    <class name="hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<std::allocator>>" persistent="false" />
             
-    <class name="hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" />
-    <class name="hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" />
-    <class name="hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" />
+    <class name="hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" persistent="false" />
+    <class name="hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" persistent="false" />
+    <class name="hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" persistent="false" />
 
     <class name="cms::cuda::Product<hcal::DigiCollection<hcal::Flavor1, calo::common::ViewStoragePolicy>>" persistent="false" />
     <class name="cms::cuda::Product<hcal::DigiCollection<hcal::Flavor3, calo::common::ViewStoragePolicy>>" persistent="false" />
@@ -26,11 +26,11 @@
     <class name="edm::Wrapper<cms::cuda::Product<hcal::DigiCollection<hcal::Flavor3, calo::common::DevStoragePolicy>>>" persistent="false" />
     <class name="edm::Wrapper<cms::cuda::Product<hcal::DigiCollection<hcal::Flavor5, calo::common::DevStoragePolicy>>>" persistent="false" />
 
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<std::allocator>>>" />
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<std::allocator>>>" />
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<std::allocator>>>" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<std::allocator>>>" persistent="false" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<std::allocator>>>" persistent="false" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<std::allocator>>>" persistent="false" />
                 
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" />
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" />
-    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor1, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" persistent="false" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor3, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" persistent="false" />
+    <class name="edm::Wrapper<hcal::DigiCollection<hcal::Flavor5, calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" persistent="false" />
 </lcgdict>

--- a/CUDADataFormats/HcalRecHitSoA/src/classes_def.xml
+++ b/CUDADataFormats/HcalRecHitSoA/src/classes_def.xml
@@ -1,9 +1,9 @@
 <lcgdict>
-    <class name="hcal::RecHitCollection<calo::common::VecStoragePolicy<std::allocator>>"/>
-    <class name="edm::Wrapper<hcal::RecHitCollection<calo::common::VecStoragePolicy<std::allocator>>>"/>
+    <class name="hcal::RecHitCollection<calo::common::VecStoragePolicy<std::allocator>>" persistent="false"/>
+    <class name="edm::Wrapper<hcal::RecHitCollection<calo::common::VecStoragePolicy<std::allocator>>>" persistent="false"/>
 
-    <class name="hcal::RecHitCollection<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>"/>
-    <class name="edm::Wrapper<hcal::RecHitCollection<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>"/>
+    <class name="hcal::RecHitCollection<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>" persistent="false"/>
+    <class name="edm::Wrapper<hcal::RecHitCollection<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>>" persistent="false"/>
 
     <class name="cms::cuda::Product<hcal::RecHitCollection<calo::common::ViewStoragePolicy>>" persistent="false" />
     <class name="edm::Wrapper<cms::cuda::Product<hcal::RecHitCollection<calo::common::ViewStoragePolicy>>>" persistent="false" />

--- a/CUDADataFormats/StdDictionaries/src/classes_def.xml
+++ b/CUDADataFormats/StdDictionaries/src/classes_def.xml
@@ -1,14 +1,14 @@
 <lcgdict>
-    <class name="std::vector<std::byte, cms::cuda::HostAllocator<std::byte, 0>>" />
+    <class name="std::vector<std::byte, cms::cuda::HostAllocator<std::byte, 0>>" persistent="false"/>
 
-    <class name="std::vector<uint8_t, cms::cuda::HostAllocator<uint8_t, 0>>" />
-    <class name="std::vector<uint16_t, cms::cuda::HostAllocator<uint16_t, 0>>" />
-    <class name="std::vector<uint32_t, cms::cuda::HostAllocator<uint32_t, 0>>" />
+    <class name="std::vector<uint8_t, cms::cuda::HostAllocator<uint8_t, 0>>" persistent="false"/>
+    <class name="std::vector<uint16_t, cms::cuda::HostAllocator<uint16_t, 0>>" persistent="false"/>
+    <class name="std::vector<uint32_t, cms::cuda::HostAllocator<uint32_t, 0>>" persistent="false"/>
 
-    <class name="std::vector<int8_t, cms::cuda::HostAllocator<int8_t, 0>>" />
-    <class name="std::vector<int16_t, cms::cuda::HostAllocator<int16_t, 0>>" />
-    <class name="std::vector<int32_t, cms::cuda::HostAllocator<int32_t, 0>>" />
+    <class name="std::vector<int8_t, cms::cuda::HostAllocator<int8_t, 0>>" persistent="false"/>
+    <class name="std::vector<int16_t, cms::cuda::HostAllocator<int16_t, 0>>" persistent="false"/>
+    <class name="std::vector<int32_t, cms::cuda::HostAllocator<int32_t, 0>>" persistent="false"/>
 
-    <class name="std::vector<float, cms::cuda::HostAllocator<float, 0>>" />
-    <class name="std::vector<double, cms::cuda::HostAllocator<double, 0>>" />
+    <class name="std::vector<float, cms::cuda::HostAllocator<float, 0>>" persistent="false"/>
+    <class name="std::vector<double, cms::cuda::HostAllocator<double, 0>>" persistent="false"/>
 </lcgdict>

